### ISSUE TITLE
Improve viewport restrictions via translateExtent & maxZoom, minZoom

### ISSUE THREAD

**Issue by peterkogo** [xyflow/xyflow#4784]
Currently, `translateExtent` is not usable in a sensible way. It is ignored while zooming and trying to actually use it for a user to reliably hide the viewport outside the specified extent is impossible.

Related issues: #4753, #4763

**Comment by bcakmakoglu** [comment]
On that note, maybe it'd make sense to be able to assign a nodeExtent that also matches the dimensions of the container?
So basically something like `nodeExtent: 'container'` (just spitballing here though), similar to 'parent'?

**Comment by peterkogo** [comment]
@bcakmakoglu How would this work with different zoom levels?

**Comment by bcakmakoglu** [comment]
Mhhhh.... good question 🤔
I was just thinking if I wanted to "fix" the translateExtent to specific bounds I'd like to be able to match the nodeExtent to that without doing some calculations but I guess using `getNodeBounds` would be enough?

**Comment by peterkogo** [comment]
Hm, I have to think about this. I see the utility in restricting the nodes to the restricted viewport. Maybe the answer would be to just pass the same extent to both `nodeExtent` and `translateExtent` (or whatever future prop will be called)

**Comment by peterkogo** [comment]
Ok, hold up I think I just figured this out. Was a little confused about what translateExtent was doing exactly. 

@bcakmakoglu  Can you please check if this is actually what we are trying to achieve here? 

https://codesandbox.io/p/sandbox/amazing-kare-jnmt6t

**Comment by bcakmakoglu** [comment]
Mh... what exactly is supposed to happen in the sandbox?
Because from what I can tell the nodes just do some weird jumps when I reach the border of the viewport and the viewport movement on drag would start. 😅

**Comment by peterkogo** [comment]
Oh wait, I forgot to fork the sandbox again.... Give me some time it might have been wrong any way. Sorry 😅

**Comment by bsvetlik** [comment]
@peterkogo I've been exploring your sandbox example and it's looking promising!
To further elevate the user experience, I'd suggest focusing on minimizing unexpected viewport jumps.
Here are a couple of scenarios where these jumps occur:
1. Drag the "with React Flow" node downwards, then back up. Click and drag the canvas.
Result: The viewport unexpectedly shifts downwards.
2. Drag the "drag me!" node outside the canvas to the left, then back to the right canvas edge.
Result: The viewport abruptly moves to the left.

**Comment by peterkogo** [comment]
https://codesandbox.io/p/sandbox/solitary-bush-9zpvhn

I had a look in the morning with a fresh pair of eyes and saw that you could improve it by making the boundary condition per axis. I think some jerking is expected to happen and it also does in gojs. 

I think some of it comes down to the fact the translate extent is applied a tick later than the auto panning.

**Comment by bsvetlik** [comment]
@peterkogo 
GoJS takes a different approach: the canvas remains fixed in place until a node is dragged beyond its edges and released. Once released, the canvas animates back to the node, providing a more controlled and predictable user experience.
See https://gojs.net/latest/samples/scrollModes.html

**Comment by peterkogo** [comment]
I disagree. Not seeing where you are dragging a node does not make any sense imo. Also, dragging the node far enough out of the view it is not sensibly able to animate to the position of the node.

I don't think this is an acceptable behavior we'd like to imitate.

https://github.com/user-attachments/assets/fdcf4fba-8c4d-4fd3-ae2f-9a1248260f8f

**Comment by bsvetlik** [comment]
@peterkogo
You're absolutely right that GoJS approach has its own drawbacks. Its advantage is smooth node and canvas movement. If you could achieve this it would be great!
